File: dsa/stack and queue/queue/circular_queue.cpp
/* TOPIC: Circular Queue */

// REFERENCE: https://youtu.be/W7uB9-TKfTg

#include <bits/stdc++.h>
using namespace std;

class CircularQueue {

    int* arr;
    int front;
    int rear;
    int size;

    public:
    CircularQueue(int size) {
        this->size = size;
        this->front = -1;
        this->rear = -1;
        this->arr = new int[this->size];
    }

    void enqueue(int val) {
        if ((front == 0 and rear == size-1) or (rear == (front-1)%(size-1))) {
            cout<<"queue is full"<<endl;
            return;
        }
        if (front == -1) {
            front = 0;
            rear = 0;
            arr[rear] = val;
            return;
        }
        rear = (rear+1)%size;
        arr[rear] = val;
        return;
    }

    void dequeue() {
        if (front == -1) {
            cout<<"queue is empty"<<endl;
            return;
        }
        if (front == rear) {
            arr[front] = -1;
            front = -1;
            rear = -1;
            return;
        }
        arr[front] = -1;
        front = (front+1)%size;
        return;
    }

    int frontVal() {
        if (front == -1) {
            return -1;
        }
        return this->arr[this->front];
    }

    bool isEmpty() {
        return front == -1;
    }

    void print() {
        for (int i=0; i<size; i++) {
            cout<<arr[i]<<" ";
        } cout<<endl;
    }

};

int main() {

    CircularQueue q(5);

    q.enqueue(1);
    q.enqueue(2);
    q.enqueue(3);
    q.enqueue(4);
    q.enqueue(5);
    q.print();
    q.dequeue();
    q.print();

    q.enqueue(6);
    q.print();
    q.dequeue();
    q.print();

    cout << q.frontVal() << endl;

    cout << q.isEmpty() << endl;

    return 0;
}

File: dsa/queue/circular_queue.py
""" Circular Queue"""

class Queue:
    def __init__(self, max_size):
        self.items = max_size * [None]
        self.max_size = max_size
        self.front = -1
        self.rear = -1

    def __str__(self):
        values = [str(item) for item in self.items if item is not None]
        return "[" + " ".join(values) + "]"

    def isFull(self):
        if self.rear + 1 == self.front:
            return True
        elif self.front == 0 and self.rear == self.max_size-1:
            return True
        else:
            return False

    def isEmpty(self):
        if self.rear == -1:
            return True
        else:
            return False

    def enqueue(self, value):
        if self.isFull():
            print("The queue is Full")
        else:
            if self.rear == self.max_size-1:
                self.rear = 0
            else:
                self.rear += 1
                if self.front == -1:
                    self.front = 0
            self.items[self.rear] = value

    def dequeue(self):
        if self.isEmpty():
            print("The queue is empty")
        else:
            first_element = self.items[self.front]
            front = self.front
            if self.front == self.rear:
                self.front = -1
                self.rear = -1
            elif self.front == self.max_size-1:
                self.front = 0
            else:
                self.front += 1
            self.items[front] = None
            return first_element
        
    def peek(self):
        if self.isEmpty():
            print("The queue is empty")
        else:
            return self.items[self.front]

    def delete(self):
        self.items = self.max_size * [None]
        self.front = -1
        self.rear = -1


q = Queue(6)
q.enqueue(1)
q.enqueue(2)
q.enqueue(3)
q.enqueue(4)
print(q)

q.dequeue()
print(q)

print(q.peek())

q.delete()
